### ISSUE TITLE
64 player scoreboard

### ISSUE THREAD

**Issue by fokkonaut** [teeworlds/teeworlds#1828]
imo we need to do something about it, it doesnt look good. only some players are listed and the rest is not visible (and x other player...). what about using a big scoreboard like this: https://ddnet.tw/client/scoreboard.png

**Comment by Dune-jr** [comment]
Ouch, this scoreboard is dreadful to look at.

I think what we currently have is best until we find something that actually looks fine with this many players (scrollbar, ...).

**Comment by 12pm** [comment]
Better a bad locking scoreboard than one that doesn’t display all players. Functionality matters most here

**Comment by Dune-jr** [comment]
If you want to push a feature in Teeworlds, it should be good-looking from the get go, not "let's add this functionality then make it look okay later somehow".

**Comment by ChillerDragon** [comment]
It looks pretty okay already? Wait what xd? We have a scrollbar in the scoreboard? Ouch rip weapon switching while looking at the scoreboard ._.

**Comment by 12pm** [comment]
Yes, but with that also comes that it should be functional, which isn't the case

**Comment by Dune-jr** [comment]
> It looks pretty okay already? Wait what xd? We have a scrollbar in the scoreboard? Ouch rip weapon switching while looking at the scoreboard ._.

No we don't.

**Comment by oy** [comment]
Agree with @Dune-jr that proposed scoreboard doesn't look good, it is cluttered and pretty much impractical. 
The information you search for should be accessable easily. It should provide information about yourself and classify it accordingly and should filter out not so relevant information. The actual scoreboard is the best for now, till someone comes up with a better one.

**Comment by fokkonaut** [comment]
Lol, "the information you search for should be accessable easily." You just outplayed yourself. Because right now you cant even see any Information xD tell me one game where the scoreboard is saying "and x other players..." this is rediculous :D I want to see every tees information on a scoreboard. 

What about making 2 scoreboard next to each other. Then we can have 16 players on each site and have 32 in total. Then we somehow need a bind or stuff to go to another site where the other 32 people are. the  the readability is still good and we have everyones information.

**Comment by nheir** [comment]
> You just outplayed yourself.

> It should provide information about yourself and classify it accordingly.

↑ the point

I'm not satisfied either by the scoreboard, still, it supports 64p without changing the size of UI and the information it gives are relevant for the player on himself.

The flaws are the hidden informations. For instance,
 - What is the score of the player that is trying to kill me ?  Is this a threat ?
 - someone starts a vote for kicking a -10 scored player but I can't check since I don't see him in the scoreboard without spectating him

> What about making 2 scoreboard next to each other.

Assuming that it is possible to have 64 players in team red while 0 in team blue, you need 4 sites.
I also assume you like seeing scoreboard while playing, at least I do, then playing, pressing +scoreboard and using binds to change sites seems really hard.
Better have a full scoreboard from the start, but it should be readable for the minimal resolution

How does the ddnet scoreboard looks like on small resolution when 64 players ? and what about 2 team games ?

**Comment by fokkonaut** [comment]
its a problem in every scenario lol, do you even play teeworlds? the best solution would be a button like you have created in the browser to expand the filter stuff but on the right side of a 32 player scoreboard to see the other 32 players. then you could hook, fire like you want but if you press on the button the scoreboard will swap

**Comment by Dune-jr** [comment]
> > I think there is a simple solution: a scrollbar that you can only left-click (no mousewheel).
>
> its a problem in every scenario lol, do you even play teeworlds? the best solution would be a button like you have created in the browser to expand the filter stuff but on the right side of a 32 player scoreboard to see the other 32 players. then you could hook, fire like you want but if you press on the button the scoreboard will swap

You mean, like a button that you can left-click? ;-)

**Comment by Stitch626** [comment]
I personally don't like that crappy ddnet-like 64p scoreboard. Thats why I've implemented a scoreboard with pages instead of a scrollbar. That would allow (active) ingame scoreboards while using different hotkeys to scroll there. No problems at all.

**Comment by 1stDecadeTW** [comment]
@Stitch626 Could you please upload a screenshot of the scoreboard you mention?

**Comment by sepulzera** [comment]
@fokkonaut What will it look like for team game modes with no teambalance and all players joined one team?

My two cents:
Currently only 16p ingame are supported and we have designed a pretty solid and beautiful scoreboard for it.
You won't ever get 64p into the scoreboard, or it will look awful like ddnet proves.

If really needed, we should implement a second, more compact scoreboard for 64p support, making it optionally. Eventually remove some information then, to make all players fit.
But, imho, we shouldn't use our limited manpower for this right now.

**Comment by Stitch626** [comment]
@1stDecadeTW This is on ddnet right now, but also works on other mods and vanilla itself.
![screenshot_2018-12-12_13-30-49](https://user-images.githubusercontent.com/442533/49870035-71634e00-fe12-11e8-969e-e6ecd3ad680d.png)
![screenshot_2018-12-12_13-30-52](https://user-images.githubusercontent.com/442533/49870036-71634e00-fe12-11e8-95eb-9951ef58f7c8.png)

**Comment by ChillerDragon** [comment]
Well I personally really like the ddnet scoreboard. Probably because I am used to it. But the problem with pages and scrolling is that I already see me switching pages forward and backward because i forgot what i saw on the previous one. And also giving the scoreboard interaction more attention than the ingame action :c

**Comment by sepulzera** [comment]
This should be tackled when implementing the (or any) race gamemode, see #1828

**Comment by chachad** [comment]
I'd like to see @Stitch626's scoreboard implementation. There is only two options I think: do it like DDNet did or use pages. Also would be good to add binds for fast page change. Like Left/Right or Up/Down buttons. Or use shift/space to move on next page. And don't save last opened page, so there will be only one direction of moving around pages. When player try to go next page after the latest, just render first page again.

**Comment by ChillerDragon** [comment]
@chachad Left/Right is for moving. As far as i know most people use A and D but Left/Right is binded default as well if I remember correctly. Shift is for emotes and space for jumping. And scrolling is for weaponswitch. Anyways all the non gameplay keys might be used by other custom binds. And in 0.6 the game was still playable with the scoreboard open which was very nice and usefull sometimes. Please don't kill that feature :c

**Comment by Dune-jr** [comment]
Default is A/D, but yeah, I agree it is not a great ide to hardcode arrows

**Comment by ChillerDragon** [comment]
> Edit: Whats with the page up/page down keys?

@Stitch626 pressing fn + up/down on a laptop can be very annoying for this task. Checking the scoreboard should be fast and easy.

**Comment by Stitch626** [comment]
I meant those keys as a good default bind...

**Comment by Dune-jr** [comment]
Bonus idea:

![image](https://user-images.githubusercontent.com/355114/67639916-b2126b80-f8ed-11e9-95d3-94e091452c9e.png)

The statboard is too narrow, but maybe that could help in the scoreboard.
Probably better if the tees prefix the text

**Comment by oy** [comment]
@Dune-jr good idea 👍

**Comment by Learath2** [comment]
I disagree heavily with the DDNet scoreboard being "dreadful", it has stood the test of time, 64p has been a thing long before vanilla caught up. If people were having trouble with any aspect of it we'd have heard of it by now given we've had all kinds of players on all kinds of aspect ratios / resolutions.

Skins help cut through the clutter to identify the person you need. TBF I can't quite imagine an interface that is not cluttered and can fit 64 players in it. Only ways of hiding information from the player behind pages which again is not convenient to navigate while playing.

**Comment by Dune-jr** [comment]
@Learath2 The look of the scoreboard is absolutely up to everyone's taste, but it is cramped for sure, and it'd be very challenging to fit everything in a screen, especially with the new scoreboard infos.

Like you say, I can't quite imagine an interface that fits 64 players without being cluttered, in fact that was the strong argument *against* the extension to 64 players for 0.7.1.

If you scroll up, I feel like we agreed on having a scoreboard on several pages, and were debating how to move, probably with pageup/down-like binds, as we want people to use the mouse freely while tabbing.

> If people were having trouble with any aspect of it we'd have heard of it by now given we've had all kinds of players on all kinds of aspect ratios / resolutions.

Tbh, even Teeworlds very rarely gets UI complaints, so I am not too surprised DDNet didn't. If you go up the issue, you will see people like @Stitch626 were unhappy about it and went to implement their own with scrollbars etc.

**Comment by Learath2** [comment]
>Like you say, I can't quite imagine an interface that fits 64 players without being cluttered, in fact that was the strong argument against the extension to 64 players for 0.7.1.

This was always a moot point, the UI is only a concern to you if you allow 64p, vanilla only supports 64p with most players being in spec.

> Tbh, even Teeworlds very rarely gets UI complaints, so I am not too surprised DDNet didn't. If you go up the issue, you will see people like @Stitch626 were unhappy about it and went to implement their own with scrollbars etc.

If there was actual trouble reading the scoreboard, there'd definitely be complaints, what are people supposed to do, just not use the scoreboard ever?

**Comment by Dune-jr** [comment]
> > Like you say, I can't quite imagine an interface that fits 64 players without being cluttered, in fact that was the strong argument against the extension to 64 players for 0.7.1.
> 
> This was always a moot point, the UI is only a concern to you if you allow 64p, vanilla only supports 64p with most players being in spec.

That's why we went with a temporary workaround scoreboard, right? Vanilla servers do not support 64p, but the UI should not be breakable by mods.

> If there was actual trouble reading the scoreboard, there'd definitely be complaints, what are people supposed to do, just not use the scoreboard ever?

Right, so you have a point there, this proves the DDNet scoreboard is reasonably readable for everyone.

. 

I would echo @chachacha3's point here: 
> There is only two options I think: do it like DDNet did or use pages. 

From there on, it is subjective.

**Comment by fokkonaut** [comment]
I also added the 64p scoreboard to F-Client, in 0.7 style. Anyone can try it out.

**Comment by ChillerDragon** [comment]
![image](https://user-images.githubusercontent.com/20344300/77768631-08253f80-7043-11ea-8880-81bcb55b8a97.png)

Its not full server but it would not get smaller and can fit in 64 in game players in this size. Imo looks fine.

**Comment by Zatline** [comment]
Skins are already falling apart in a soup of unsharpness, and I assume that's at a regular resolution.

**Comment by Sonix-** [comment]
Would increasing the scoreboard's size to something like this make it more readable? (Red outline area)
<img width="1440" alt="64pscoreboard" src="https://user-images.githubusercontent.com/2030615/79150376-f4126980-7dc8-11ea-8642-480f0909c901.png">

Font size could be increased since it allows more space.
And eventually add a bottom bar to display score limit etc.

**Comment by ChillerDragon** [comment]
> Skins are already falling apart in a soup of unsharpness, and I assume that's at a regular resolution.

Yes that is 1920x1080

>  The spectators box should be removed only if more than n. players, then moved to the side.

It would be nice to not loose any information or tee.

**Comment by TsFreddie** [comment]
In CSGO, you can press Alt while holding Tab to bring up the cursor. We can do the same idea to enable UI elements like scrollbar, and even integrate some of the feature in Players tab (and maybe statboard?) into the scoreboard directly.

**Comment by sepulzera** [comment]
> In CSGO, you can press Alt while holding Tab to bring up the cursor.

Is that true? I played CSGO for quite some hours, but never knew this. #Usabilty
